Permutative.agda

```diff
@@ -0,0 +1,85 @@
+{-# OPTIONS --type-in-type --without-K #-}
+
+module Permutative where
+
+open import Agda.Primitive
+
+module Cat where
+  record T : Set where
+    no-eta-equality
+    field
+      obj
+        : Set
+      hom
+        : (a b : obj) → Set
+      idn
+        : ∀ {a}
+        → hom a a
+      cmp
+        : ∀ {a b c}
+        → (g : hom b c)
+        → (f : hom a b)
+        → hom a c
+  open T public
+
+  _ᵒᵖ : (𝒞 : T) → T
+  𝒞 ᵒᵖ = record
+    { obj = obj 𝒞
+    ; hom = λ a b → hom 𝒞 b a
+    ; idn = λ {a} → idn 𝒞 {a}
+    ; cmp = λ g f → cmp 𝒞 f g
+    }
+
+module ℕ where
+  data T : Set where
+    z : T
+    s_ : (n : T) → T
+
+module Dir where
+  data T : Set where
+    #0 #1 : T
+
+module Symbol where
+  data T : Set where
+    2+_ : ℕ.T → T
+
+module □ where
+  infixr 0 _∷_
+  data Ctx : Set where
+    [] : Ctx
+    _∷_ : (x : Symbol.T) (xs : Ctx) → Ctx
+
+  module ⊆ where
+    data _T_ : (Δ Γ : Ctx) → Set where
+      wkn : ∀ {A Δ Γ} → Δ T Γ → Δ T (A ∷ Γ)
+      xch : ∀ A₀ A₁ {Δ} → (A₀ ∷ A₁ ∷ Δ) T (A₁ ∷ A₀ ∷ Δ)
+      idn : ∀ {Δ} → Δ T Δ
+      cmp : ∀ {Δ Γ Θ} → Γ T Θ → Δ T Γ → Δ T Θ
+
+  data Hom (Δ Γ : Ctx) : Set where
+    inj : Δ ⊆.T Γ → Hom Δ Γ
+    dir : Dir.T → Hom Δ Γ
```

Comment:
I think so. The idea with `⊆` is that it encodes injections as "permutations with an enlarged codomain", where the latter part comes from `wkn`, which lets you skip over certain elements. You should be able to interpret this data type into a function which would be closer to the original version. I just noticed that it's missing a constructor for `cons` but that's easy to fix. With that, it's also quite similar to a type for explicit substitutions. The `Hom` thing should satisfy `I -> J + 2` under the equivalence `(I -> J) + (I -> 2)` which then reduces to `(I -> J) + 2` (thinking of `I -> 2` as a generalized element of `2`).